src/LuaUtils.cpp

```diff
@@ -176,122 +176,261 @@ static bool _import_core(lua_State *L, const std::string &importName)
-		lua_getinfo(L, "S", &ar);
+	bool loaded = false;
+	RefCountedPtr<FileSystem::FileData> fileData = FileSystem::gameDataFiles.ReadFile(name);
 
-		if (ar.source) {
-			int start = 0;
-			int end = strlen(ar.source);
+	// if file exists
+	if (fileData)
+	{
+		pi_lua_dofile(L, *fileData, 1);
+		loaded = true;
+	}
 
-			if (ar.source[start] != '@' && end > 5)
-				start = 4;
+	LUA_DEBUG_END(L, 0);
+	return loaded;
+}
 
-			if (ar.source[start] == '@') {
-				start++;
+/**
+* Import or load from cache lua module.
+* Returns true if module imported, puts return of module to lua stack.
+* If module not loaded returns false, puts error message to lua stack.
+* If isFullName disabled tries to check with .lua extension and also
+* relative by importDirectories. Else checks only by importName.
+*/
+static bool _import(lua_State *L, const std::string &importName, bool isFullName = false)
+{
+	LUA_DEBUG_START(L);
 
-				static char source[1024];
-				strncpy(source, &ar.source[start], end-start+1);
-				DEBUG_PRINTF("import: called from %s\n", source);
+	bool isImported = false;
+	bool wasCached = false;
+	std::string fileName;
+	const std::string& callerDirectory = get_caller_directory(L);
 
-				for (end = end-start-1; end >= 0; end--)
-					if (source[end] == '/') {
-						source[end] = '\0';
-						break;
-					}
+	// contans list of paths where file can be imported
+	std::vector<std::string> importDirectories =
+	{
+		"libs"
+	};
 
-				DEBUG_PRINTF("import: relative path %s\n", source);
+	// add caller directory to source if exists
+	if (!isFullName && !callerDirectory.empty())
+	{
+		DEBUG_PRINTF("import [%s]: called from %s\n", importName.c_str(), callerDirectory.c_str());
 
-				cacheName = FileSystem::JoinPath(source, importName+".lua");
-				lua_getfield(L, -1, cacheName.c_str());
-				if (!lua_isnil(L, -1)) {
-					// found it, use it
-					DEBUG_PRINTF("import: found cached %s\n", cacheName.c_str());
-					lua_remove(L, -2);
-					return 1;
-				}
-				lua_pop(L, 1);
-				DEBUG_PRINTF("import: not cached %s\n", cacheName.c_str());
+		// check if caller not exists in import directory
+		bool directoryFound = false;
+		for (auto it = importDirectories.begin(); it != importDirectories.end(); ++it)
+		{
+			if (*it == callerDirectory)
+			{
+				directoryFound = true;
+				break;
+			}
+		}
+
+		if (!directoryFound)
+			importDirectories.push_back(callerDirectory);
+	}
+
+
+	// first check the cache (its there if it was previously imported from core)
+	{
+		DEBUG_PRINTF("import [%s]: trying to load from cache\n", importName.c_str());
+
+		// load cache
+		lua_getfield(L, LUA_REGISTRYINDEX, "Imports");
+		assert(lua_istable(L, -1));
 
-				code = FileSystem::gameDataFiles.ReadFile(cacheName);
+		// check original name, with lua extension and relative by dirs by same rule
+		// if assumptions disabled, loop pass once and check only original importName
+		std::string cacheName;
+		for (int i = 0; i < (importDirectories.size() * !isFullName + 1); i++)
```

Comment:
While the _import function is far too long as is, I'm pretty sure I'd rather have a more explicit way of expressing the boundary check here. Also, using a boolean operation result directly in an arithmetic expression is really not a good idea in my book.
```cpp
size_t dirsToCheck = 1;
if (!isFullName)
    dirsToCheck += importDirectories.size();
for (int i = 0; i < dirsToCheck; ++i)
```

Reply:
Agreed, I will change it